stan/math/rev/functor/cvodes_integrator_adjoint.hpp

```diff
@@ -0,0 +1,829 @@
+
+          int error_code
+              = CVodeF(solver_->cvodes_mem_, t_final,
+                       solver_->nv_state_forward_, &t_init, CV_NORMAL, &ncheck);
+
+          if (error_code == CV_TOO_MUCH_WORK) {
+            throw_domain_error(solver_->function_name_, "", t_final,
+                               "Failed to integrate to next output time (",
+                               ") in less than max_num_steps steps");
+          } else {
+            check_flag_sundials(error_code, "CVodeF");
+          }
+
+        } else {
+          int error_code
+              = CVode(solver_->cvodes_mem_, t_final, solver_->nv_state_forward_,
+                      &t_init, CV_NORMAL);
+
+          if (error_code == CV_TOO_MUCH_WORK) {
+            throw_domain_error(solver_->function_name_, "", t_final,
+                               "Failed to integrate to next output time (",
+                               ") in less than max_num_steps steps");
+          } else {
+            check_flag_sundials(error_code, "CVode");
+          }
+        }
+      }
+      store_state(n, state_forward_, y_return[n]);
+
+      t_init = t_final;
+    }
+
+    forward_returned_ = true;
+    return y_return;
+  }
+
+  virtual void chain() {
+    if (solver_->cvodes_mem_ == nullptr) {
+      return;
+    }
+
+    if (!forward_returned_) {
+      return;
+    }
+
+    if (!is_var_return) {
+      return;
+    }
+
+    // for sensitivities wrt to ts we do not need to run the backward
+    // integration
+    if (is_var_ts) {
+      for (int i = 0; i < ts_.size(); ++i) {
+        Eigen::VectorXd step_sens = Eigen::VectorXd::Zero(N_);
+        for (int j = 0; j < N_; ++j) {
+          step_sens.coeffRef(j) += non_chaining_varis_[i * N_ + j]->adj_;
+        }
+
+        ts_varis_[i]->adj_ += step_sens.dot(
+            rhs(value_of(ts_[i]), solver_->y_[i], value_of_args_tuple_));
+      }
+
+      if (is_var_only_ts) {
+        return;
+      }
+    }
+
+    state_backward_.setZero();
+    quad_.setZero();
+
+    // At every time step, collect the adjoints from the output
+    // variables and re-initialize the solver
+    double t_init = value_of(ts_.back());
+    for (int i = ts_.size() - 1; i >= 0; --i) {
+      // Take in the adjoints from all the output variables at this point
+      // in time
+      for (int j = 0; j < N_; ++j) {
+        state_backward_.coeffRef(j) += non_chaining_varis_[i * N_ + j]->adj_;
+      }
+
+      double t_final = value_of((i > 0) ? ts_[i - 1] : t0_);
+      if (t_final != t_init) {
+        if (!backward_is_initialized_) {
+          check_flag_sundials(CVodeCreateB(solver_->cvodes_mem_,
+                                           solver_backward_, &index_backward_),
+                              "CVodeCreateB");
+
+          check_flag_sundials(
+              CVodeSetUserDataB(solver_->cvodes_mem_, index_backward_,
+                                reinterpret_cast<void*>(this)),
+              "CVodeSetUserDataB");
```

Comment:
Should we be doing reinterpret casts or static casts when going to and fro from `void*`? I want to say reinterpret cast, which if so we should also do in the ODE testing suite

Reply:
I did your other very good suggestions.